### ISSUE TITLE
Compiler crashes when using @ptrCast between optional pointers to zero-sized and non-zero-sized types

### ISSUE THREAD

**Issue by dominikh** [ziglang/zig#6861]
Running 81f5e1e66756096b101f9e30e7f41a24eeaa79c5, the following two code samples lead to the following crashes in the compiler:

```
fn init(data: ?*u1) ?*u0 {
    return @ptrCast(?*u0, data);
}

pub fn main() !void {
    _ = init(null);
}
```

```
$ zig build-exe bug.zig
Assertion failed at /build/source/src/stage1/analyze.cpp:619 in get_pointer_to_type_extra2. This is a bug in the Zig compiler.
Unable to dump stack trace: debug info stripped
zsh: abort (core dumped)  zig build-exe bug.zig
```

```
fn init(data: ?*u0) ?*u1 {
    return @ptrCast(?*u1, data);
}

pub fn main() !void {
    _ = init(null);
}
```

```
$ zig build-exe bug.zig
broken LLVM module found: Invalid bitcast
  %2 = bitcast i1 %1 to i1*, !dbg !17234

This is a bug in the Zig compiler.
Unable to dump stack trace: debug info stripped
zsh: abort (core dumped)  zig build-exe bug.zig
```

**Comment by theInkSquid** [comment]
I think I found the source of the bug in `ir_analyze_ptr_cast`, and It made me think about some other cases:

should this be allowed?

```rust
var x: u0 = 0;
const x_ptr: *u0 = &x;

_ = @ptrCast(?*void, x_ptr);
```

It's a cast from a zero-size-type to a non-zero-size-type, but even though `?*void` has bits, the pointer part doesn't. Should this be allowed?

**Comment by mikdusan** [comment]
with latest compiler:
- compiler no longer core dumps
- appears this was fixed between 0.9.1 and 0.10.0

<details>
<summary>repro.zig</summary>

```zig
fn init(data: ?*u1) ?*u0 {
    return @ptrCast(?*u0, data);
}

pub fn main() !void {
    _ = init(null);
}
```
</details>

<details>
<summary>repro2.zig</summary>

```zig
export fn foo() void {
    var x: u0 = 0;
    const x_ptr: *u0 = &x;

    _ = @ptrCast(?*void, x_ptr);
}
```
</details>

```sh
$ uname -a
Linux archlinux 6.1.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Sat, 31 Dec 2022 17:40:35 +0000 x86_64 GNU/Linux

$ zig version
0.11.0-dev.1247+87b223428

$ zig build-obj repro.zig
# no stdout/stderr (good)

$ zig build-obj repro2.zig
# no stdout/stderr (good)
```
